awscli/customizations/history/db.py

```diff
@@ -0,0 +1,256 @@
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _encode_mutable_mapping(self, obj):
+        return dict(obj)
+
+    def _encode_binary(self, obj):
+        # On PY3 a bytes like object that contains entirely valid utf-8 bytes
+        # will be routed here instead of the encode method. In the case of
+        # utf-8 bytes we want to make the stored data as readble as possible so
+        # first we try to decode it as utf-8, if that fails we will fall back
+        # to using a base64 encoded string.
+        try:
+            encoded = obj.decode('utf-8')
+        except UnicodeDecodeError:
+            encoded = ensure_text_type(base64.b64encode(obj))
+        return encoded
+
+    def _encode_datetime(self, obj):
+        return obj.isoformat()
+
+    def _unknown(self, obj, type_name):
+        return type_name
+
+    def encode(self, obj):
+        # The default method is not called in PY2 where the JSONEncoder thinks
+        # it can handle a bytes object because it can't tell it apart from a
+        # str object. In PY3 this is handled by the _process_bytes method.
+        # For PY2 we have to override the encode method where the encoding of
+        # a bytes-like string will fail, catch the UnicodeDecodeError and
+        # then call _process_bytes and retry the encoding.
+        try:
+            encoded = super(PayloadSerializer, self).encode(obj)
+            return encoded
+        except UnicodeDecodeError:
+            b64encoded = self._process_bytes(obj)
+            encoded = self.encode(b64encoded)
+            return encoded
+
+    def default(self, obj):
+        if isinstance(obj, binary_type):
+            return self._encode_binary(obj)
+        elif isinstance(obj, datetime.datetime):
+            return self._encode_datetime(obj)
+        elif isinstance(obj, MutableMapping):
+            return self._encode_mutable_mapping(obj)
+        else:
+            return type(obj).__name__
+
+
+class DatabaseRecordWriter(object):
+    _WRITE_RECORD = """
+    INSERT INTO records(id, request_id, source, event_type, timestamp, payload)
+    VALUES (?,?,?,?,?,?)"""
+    _REQUEST_LIFECYCLE_EVENTS = set(
+        ['API_CALL', 'HTTP_REQUEST', 'HTTP_RESPONSE', 'PARSED_RESPONSE'])
+    _START_OF_REQUEST_LIFECYCLE_EVENT = 'API_CALL'
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._request_ids = {}
+        self._identifier = None
+
+    def write_record(self, record):
+        # This method is not threadsafe by itself, it is only threadsafe when
+        # used inside a handler bound to the HistoryRecorder in botocore which
+        # is protected by a lock.
+        cursor = self._connection.cursor()
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
```

Comment:
This seems like the wrong level of abstraction to track this at.  I wouldn't expect a database record writer to know anything about request lifecycle events.

I would expect:

* Handler - bridge between history records and the backend storage
* DatabaseWriter - munges the records so they can be stored in the database (payload serialization, json encoding, etc).  Coordinate with:
* DBConnection - low level details of talking to sqlite3, dealing with transactions, etc.

Reply:
My thought here was this is actually somewhat a property of the database given that it has its own column, and so the record writer that writes this particular schema already has to have knowledge about it separate from the simple write payload in a row logic. I could do this logic in botocore and put it as part of the payload.